src/web-pathing.md

```diff
@@ -0,0 +1,143 @@
+that ensures compatibility with existing HTTP and Web Platform standards.
+
+## Introduction
+
+TODO: Clearly explain why the specification exists, what is the problem solved here.
+
+This document specifies details of pathing for content paths that start with
+`/ipfs` and `/ipns` namespaces, and why a logical content root included in a
+content path can facilitate security isolation and relative pathing in web
+contexts.
+
+Specification includes guidance around aspects such as hash functions,
+multibases, CID versions, codecs, and how they impact implementation's ability
+to translate pathing into traversal of a DAG.
+
+The goal of this specification is to enable competing and interoperable
+implementations, all while ensuring seamless traversal of paths within the web
+ecosystem.
+
+## Specification
+
+TODO: Explain things in depth.
+The resulting specification should be detailed enough to allow competing,
+interoperable implementations.
+
+### TODO: things to cover
+
+- TODO: why it's called "web pathing": ensuring pathing is interoperable with how existing http and web platform works; covers both /ipfs and /ipns namespace semantics; defines logical content root CID that can be mapped to URL / root which enables subdomain/dnslink gateways and ipfs:// and ipns:// protocol handlers to load existing datasets, websites, and assets with relative pathing without the need for modifying them; 
+
+- TODO: how web pathing is applied to CLI Tools; path gateways; and origin contexts: subdomain/dnslink, ipfs:// ipns:// URIs
+
+- TODO: MUSTs, SHOULDs and MAYs in relation to
+
+  - TODO: multihash functions
+    - MUSTs
+      - `sha2-256` (`0x12`)
+      - `blake2b-256` (`0xb220`)
+      - `blake3` (`0x1e`)
+      - `identity` (`0x00`) (i.e. the data itself inlined in place of a hash)
+          - TODO: Identity CIDs MUST NOT generate network I/O such as bitswap, http request, since the data is always available in Multihash itself
+    - SHOULDs
+      - `sha2-384` (`0x20`, aka SHA-384; as specified by [FIPS 180-4](https://csrc.nist.gov/pubs/fips/180-4/upd1/final)) TODO: where is this used? why is this on the list?
+      - sha3-512 TODO: code for such label does not exist, a typo in prior notes? follow up required
+
+  - TODO: mutlibases
+    - MUSTs
+      * f - base16
+      * b - base32
+      * k - base36
+      * z - base58btc (case-sensitive!)
+      * u - base64url (case-sensitive!)
+    - SHOULDs
+      * F - base16 (uppercase)
+      * B - base32 (uppercase)
+      * K - base36 (uppercase)
+
+  - TODO: cid versions
+    - MUST:
+      - CIDv1 (`0x01`)
+      - CIDV0 (Multihash encoded with `base58btc`, with implicit dag-pb `0x70` codec)
+
+  - TODO: multicodecs that are required to facilitate path traversal
+    - DAG-PB
+    - RAW
+    - libp2p-key (for IPNS names)
+    - DAG-CBOR
+    - DAG-JSON
+
+- TODO: MUST support UnixFS pathing
+  - TODO: traversing HAMTs
+  - TODO: traversing symlinks
```

Comment:
I have questions, and not sure if I should be instead commenting [here](https://github.com/ipfs/specs/issues/368) ?

There's a few basic forms I could imagine this working in, and they're not necessarily incompatible:

* /ipfs/cid1/a = "/ipfs/cid2/c" : /ipfs/cid1/a/b -> /ipfs/cid2/c/b 
  - Replace all left of and including current path element with link contents.
  - IIRC I believe the gateway conformance test has this, so I'm guessing this is the real thing.
  - Are we allowed to link to /ipns/ namespace?
  - If so... even DNSLink? The link would still be immutable, but fully resolved what looks like part of your tree now depends on your local DNS setup?
* /ipfs/cid1/a = "c" : /ipfs/cid1/a/b -> /ipfs/cid1/c/b (i.e. not starting with /)
  - Replace current path element with link contents.
  - I read someone talking about converting tar to car and if so there's an important special case...
  - "../b" : If allowed we might need rules about this.
* /ipfs/cid1/a/b = "/c" : /ipfs/cid1/a/b -> /ipfs/cid1/c
  - Replace current path element and everything between the root and current element.
  - Need rules about DAGs that contain a directory under root named either /ipfs/ or /ipns/ etc.
  - I don't love features that break DAG symmetry, but others seem to :shrug:

Reply:
Also, how does this interact with _redirects (since it both has to be in root and its redirects can be relative to the root)?
Site A has a _redirects with splat to /a.html
Site B has a symlink (called link) to a's root, and its own redirects splat to /b.html
`ipfs://B/link/notfound.html`
becomes what exactly?

In my current [PR](https://github.com/little-bear-labs/ipfs-chromium/pull/75) it would redirect to `ipfs://B/link/a.html` (e.g. it respects A's _redirects file, and does it relative to A's root). But if A did not have a redirect, it would be not found (e.g. B's _redirects is ignored).

Feels weird.

Reply:
@John-LittleBearLabs _(I've realized we've discussed this during one of sync calls but did not reply here)_

- symlinks are generally underspecified and not used much. I would mark this as unspecified behavior in this spec until we land https://github.com/ipfs/specs/pull/331 
- that being said, if you already implemented symlink support, its ok, only caveat is that following symlink should not allow for going beyond the content root (/ipfs/cid), so `/ipfs/cid1/a` pointing at `/ipfs/cid2/b` or `../cid2/b` must error 
- rules from `_redirects` are executed only when requested content path is missing within same origin (based on root CID).  in scenario you described you operate under origin B and it is not aware of _redirects from origin A (so _redirects is not executed)